### PR TITLE
Add SummaryReportBuilder to PmdTester

### PR DESCRIPTION
**Pull request by djydewang**
The summary report [demo](https://djydewang.github.io/summary).

### CODE REVIEW

File: lib/pmdtester/builders/diff_report_builder.rb [hunk 1]
```diff
@@ -0,0 +1,220 @@
+require 'nokogiri'
+require_relative './html_report_builder'
+
+module PmdTester
+  # Building diff report according to the output of DiffBuilder
+  class DiffReportBuilder < HtmlReportBuilder
+    NO_DIFFERENCES_MESSAGE = 'No differences found!'.freeze
+
+    def build(project, report_diff)
+      project.diffs_exist = report_diff.diffs_exist?
```
Comment (adangel) [inline lib/pmdtester/builders/diff_report_builder.rb]: This looks a bit strange: Is this the correct place, to handover this information?
Maybe it's better done, before the diffreportbuilder is called?

File: lib/pmdtester/builders/summary_report_builder.rb [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    def build(projects, base_name, patch_name)
+      @projects = projects
+      @base_details = get_branch_details(base_name)
+      @patch_details = get_branch_details(patch_name)
+
+      FileUtils.mkdir_p(REPORT_DIR) unless File.directory?(REPORT_DIR)
+      index = File.new(INDEX_PATH, 'w')
+
+      html_report = build_html_report('Summary report')
+      copy_css(REPORT_DIR)
```
Comment (adangel) [inline lib/pmdtester/builders/summary_report_builder.rb]: I think, this should be optimized. We copy the css files for the summary report and for each project report.
We should be able to reuse the css files for the project reports from the summary report.

File: lib/pmdtester/builders/summary_report_builder.rb [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+      details
+    end
+
+    def build_body(doc)
+      build_branch_details_section(doc)
+      build_projects_section(doc)
+    end
+
+    def build_branch_details_section(doc)
+      doc.div(class: 'section', name: 'Branch details') do
```
Comment (adangel) [inline lib/pmdtester/builders/summary_report_builder.rb]: The name attribute should be the id. And just a personal preference of me: I tend to use only lower-case letters for the ids and avoid using any special characters and spaces.

File: lib/pmdtester/builders/summary_report_builder.rb [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+      doc.thead do
+        doc.tr do
+          doc.th 'Item'
+          doc.th 'base'
+          doc.th 'patch'
+        end
+      end
+    end
+
+    def build_branch_table_body(doc)
```
Comment (adangel) [inline lib/pmdtester/builders/summary_report_builder.rb]: I just got an idea, what could be useful: github allows to craft urls, that shows you the diff between two commits/branches, e.g. https://github.com/pmd/pmd/compare/master...adangel:issue-1186
So, in general: https://github.com/pmd/pmd/compare/#{base_branch}...#{patch_branch}
On the other hand, when we go on to integrate this in pull request builds, we probably can provide a link to the original pull request.

File: lib/pmdtester/builders/summary_report_builder.rb [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    def build_branch_table_row(doc, item, base, patch)
+      doc.tr do
+        doc.td(class: 'c') { doc.text item }
+        doc.td(class: 'a') { doc.text base }
+        doc.td(class: 'b') { doc.text patch }
+      end
+    end
+
+    def build_projects_section(doc)
+      doc.div(class: 'section', name: 'Projects') do
```
Comment (adangel) [inline lib/pmdtester/builders/summary_report_builder.rb]: name -> id

File: lib/pmdtester/builders/summary_report_builder.rb [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        build_projects_table_head(doc)
+        build_projects_table_body(doc)
+      end
+    end
+
+    def build_projects_table_head(doc)
+      doc.thead do
+        doc.tr do
+          doc.th 'project diff report link'
+          doc.th 'project name'
```
Comment (adangel) [inline lib/pmdtester/builders/summary_report_builder.rb]: I would actually remove the first column and make the project name the link to the reports.

File: test/test_summary_report_builder.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
+require 'test/unit'
+require_relative '../lib/pmdtester/builders/summary_report_builder'
+require_relative '../lib/pmdtester/parsers/projects_parser'
+
+# Unit test class for PmdTester::SummaryReportBuilder
+class TestSummaryReportBuilder < Test::Unit::TestCase
+  def test_summary_report_builder
```
Comment (adangel) [inline test/test_summary_report_builder.rb]: We probably need here a `rake clean` call, too.

File: lib/pmdtester/builders/diff_report_builder.rb (hunk 1 above)
Comment (adangel) [reply lib/pmdtester/builders/diff_report_builder.rb]: Ok, I've now seen, where it is used. First, for each project, the html report is created - and this sets the flag "diffs_exist" on the project. The flag is later on used for the summary report. So it is crucial, to render the reports of each project into html before rendering the summary... this is a flow dependency.

How about using the project object for storing the results (the report_diff object), too? That way, the summary builder can ask the project, whether a diff exists and the project in turn asks the report diff.
The project object would start out with the basic information from the project list (the configuration with which PmdTester has been started) and we would add more information with the progress.
If you rework this, I suggest to do this in a new PR.

**Comment by adangel** [comment]
@djydewang I've updated the wiki page with an example: https://github.com/pmd/pmd/wiki/%5BGSoC-2018%5D-Automated-regression-tests-against-real-world-projects#example

I've now also an account at chunk.io
here's one real report: https://chunk.io/adangel/2e9e4bc8f6b840459e3dfc1770d68ed6/index.html

I've noticed the following:

There seems to be a problem with last commit sha of the patch (it is the commit message... maybe, this happens if the patch is a git tag?).

I also noticed, that the execution time of the single projects is much better readable (00:03:31 vs. 262.89510442).

The diff exist flag seems to be not correct - it shows false on the summary, but the projects actually have some diffs...

File: lib/pmdtester/builders/diff_report_builder.rb [hunk 8]
```diff
@@ -0,0 +1,220 @@
+require 'nokogiri'
+require_relative './html_report_builder'
+
+module PmdTester
+  # Building diff report according to the output of DiffBuilder
```
[the remark this answers is not in the document]
Comment (djydewang) [reply lib/pmdtester/builders/diff_report_builder.rb]: Done.

File: lib/pmdtester/builders/summary_report_builder.rb (hunk 2 above)
Comment (djydewang) [reply lib/pmdtester/builders/summary_report_builder.rb]: Each project has its own css resource file that can make each project's diff html report independent,
which means we can upload single project diff report to somewhere, although, I don't know if there will be such a usage scenario in the future :smile:

File: lib/pmdtester/builders/summary_report_builder.rb (hunk 3 above)
Comment (djydewang) [reply lib/pmdtester/builders/summary_report_builder.rb]: Done.

File: lib/pmdtester/builders/summary_report_builder.rb (hunk 5 above)
Comment (djydewang) [reply lib/pmdtester/builders/summary_report_builder.rb]: Done.

File: lib/pmdtester/builders/summary_report_builder.rb (hunk 6 above)
Comment (djydewang) [reply lib/pmdtester/builders/summary_report_builder.rb]: Done.

File: test/test_summary_report_builder.rb (hunk 7 above)
Comment (djydewang) [reply test/test_summary_report_builder.rb]: Done.

**Comment by djydewang** [comment]
> There seems to be a problem with last commit sha of the patch (it is the commit message... maybe, this happens if the patch is a git tag?).
> 
> I also noticed, that the execution time of the single projects is much better readable (00:03:31 vs. 262.89510442).
> 
> The diff exist flag seems to be not correct - it shows false on the summary, but the projects actually have some diffs...

I'v fixed issues above. And I've noticed the following:

- links to source code of open projects are wrong. the `webview url` should be `{connection_url}/tree/{branch/tag}` rather than just equal `connection url` for those projects that are stored on github. https://github.com/pmd/pmd-regression-tester/blob/e0c9ba80e0dc22b3a99e8903322e7dfb42eba7bf/lib/pmdtester/project.rb#L21
We can fixed it by providing` webview-url` for the project or generate right `webview-url` for project when they aren't provided.

**Pull request merged**